pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c

```diff
@@ -10,48 +10,119 @@
 	__type(value, struct hist);
 } hists SEC(".maps");
 
 static struct hist zero;
 
+/*
+ * We cannot use the following:
+ * __builtin_memcmp(targ_*addr_v6, *, sizeof targ_*addr_v6);
+ * Indeed, by using the builtin, we would discard the volatile qualifier of
+ * targ_*addr_v6, so the compiler would optimize it and replaces the call
+ * with 0.
+ * So, using the volatile qualifier ensures this function is called at runtime.
+ */
+static bool inline ipv6_is_not_zero(const volatile __u8 addr[IPV6_LEN])
+{
+	for (int i = 0; i < IPV6_LEN; i++)
+		if (addr[i])
+			return true;
+	return false;
+}
+
+static bool inline ipv6_are_different(const volatile __u8 a[IPV6_LEN], const __u8 b[IPV6_LEN])
+{
+	for (int i = 0; i < IPV6_LEN; i++)
+		if (a[i] != b[i])
+			return true;
+	return false;
+}
+
 static int handle_tcp_rcv_established(struct sock *sk)
 {
 	const struct inet_sock *inet = (struct inet_sock *)(sk);
 	struct tcp_sock *ts;
 	struct hist *histp;
+	struct hist_key key = { 0 };
 	u64 slot;
-	u32 srtt, key;
+	u32 srtt;
 
 	if (targ_sport && targ_sport != BPF_CORE_READ(inet, inet_sport))
 		return 0;
+
 	if (targ_dport && targ_dport != BPF_CORE_READ(sk, __sk_common.skc_dport))
 		return 0;
-	if (targ_saddr && targ_saddr != BPF_CORE_READ(inet, inet_saddr))
-		return 0;
-	if (targ_daddr && targ_daddr != BPF_CORE_READ(sk, __sk_common.skc_daddr))
+
+	key.family = BPF_CORE_READ(sk, __sk_common.skc_family);
+	switch (key.family) {
+	case AF_INET:
+		/* If we set any of IPv6 address, we do not care about IPv4 ones. */
+		if (ipv6_is_not_zero(targ_saddr_v6) || ipv6_is_not_zero(targ_daddr_v6))
+			return 0;
+
+		if (targ_saddr && targ_saddr != BPF_CORE_READ(inet, inet_saddr))
+			return 0;
+
+		if (targ_daddr && targ_daddr != BPF_CORE_READ(sk, __sk_common.skc_daddr))
+			return 0;
+
+		break;
+	case AF_INET6:
+		/*
+		 * Reciprocal of the above: if we set any of IPv4 address, we do not care
+		 * about IPv6 ones.
+		 */
+		if (targ_saddr || targ_daddr)
+			return 0;
+
+		if (ipv6_is_not_zero(targ_saddr_v6)
+		    && ipv6_are_different(targ_saddr_v6, BPF_CORE_READ(inet, pinet6, saddr.in6_u.u6_addr8)))
+			return 0;
+
+		if (ipv6_is_not_zero(targ_daddr_v6)
+		    && ipv6_are_different(targ_daddr_v6, BPF_CORE_READ(sk, __sk_common.skc_v6_daddr.in6_u.u6_addr8)))
+			return 0;
+
+		break;
+	default:
 		return 0;
+	}
 
 	if (targ_laddr_hist)
-		key = BPF_CORE_READ(inet, inet_saddr);
+		if (key.family == AF_INET6)
+			bpf_probe_read_kernel(&key.addr, sizeof(key.addr), BPF_CORE_READ(inet, pinet6, saddr.in6_u.u6_addr8));
+		else
+			bpf_probe_read_kernel(&key.addr, sizeof(inet->inet_saddr), &inet->inet_saddr);
 	else if (targ_raddr_hist)
-		key = BPF_CORE_READ(inet, sk.__sk_common.skc_daddr);
+		if (key.family == AF_INET6)
+			bpf_probe_read_kernel(&key.addr, sizeof(key.addr), BPF_CORE_READ(sk, __sk_common.skc_v6_daddr.in6_u.u6_addr8));
+		else
+			bpf_probe_read_kernel(&key.addr, sizeof(inet->sk.__sk_common.skc_daddr), &sk->__sk_common.skc_daddr);
```

Comment:
nit: Just for consistency with the way we are accessing it:

```suggestion
			bpf_probe_read_kernel(&key.addr, sizeof(sk->__sk_common.skc_daddr), &sk->__sk_common.skc_daddr);
```